Push function pointer as CFunction

GlebChili opened the pull request.

GlebChili committed:
PushCFunction for function pointer added; GmodInterop.GetLuaFromState

GlebChili committed:
Use function pointers in GetLuaFromState

GlebChili commented on gm_dotnet_managed/Tests/PushCFuncPtr.cs:
```diff
@@ -0,0 +1,75 @@
+            {
+                unsafe
+                {
+                    lua.PushCFunction((delegate* unmanaged[Cdecl]<IntPtr, int>)(delegate* managed<IntPtr, int>)&TestFunc);
```
This horrible double type cast will go away as soon as .NET 5 SDK RC2 is available.

AaronJThompson replied:
Shall we wait for RC2 before merging this?

GlebChili replied:
Will fix as soon as RC2 is out, works for now so we can merge as it is.

GlebChili replied:
It is a test, so there will be no changes to the core codebase anyway.

GlebChili committed:
Merge pull request #50 from GmodNET/master

push master to push_function_pointer

AaronJThompson commented on gm_dotnet_managed/Tests/PushCFuncPtr.cs:
```diff
@@ -0,0 +1,75 @@
+
+                    if(ret_number != random_number)
+                    {
+                        throw new Exception("Returned number is invalid");
```
Just to confirm, is the standard styling for the project always braces even for single line bodies?

GlebChili replied:
Kind of. I always write like this, so all other parts of project are formatted in the same manner. So yes, we can consider it as a part of project's code style.

AaronJThompson commented on gm_dotnet_managed/GmodNET/Startup.cs:
```diff
@@ -156,7 +156,10 @@ internal static IntPtr Main(IntPtr lua_base, IntPtr native_version_string, int v
 
                     LuaInterop.push_c_function_safe = CreateNativeCaller<Action<IntPtr, IntPtr, IntPtr>>(params_from_native_code[54]);
+
+                    GmodInterop.lua_extructor = &LuaInterop.ExtructLua;
```
Wait, do you mean Extract rather than Extruct?

GlebChili replied:
Oops, will fix it now 😂

AaronJThompson commented on gm_dotnet_managed/GmodNET/Lua.cs:
```diff
@@ -367,11 +368,28 @@ public void PushCFunction(CFuncManagedDelegate managed_function, bool use_safe_e
+                throw new ArgumentNullException("function_pointer", "Parameter can't be nullptr.");
+            }
+
+            push_c_function(ptr, (IntPtr)function_pointer);
```
Use local var int_ptr instead of casting again

GlebChili replied:
Fixed

GlebChili committed:
PR fix

Fix extractor typo (was extrUctor)

Do not cast the same pointer to IntPtr twice

GlebChili commented:
@AaronJThompson Take a look again please.

AaronJThompson reviewed (approved):
Looks good. Approved.
As a side note, it would be awesome if we could add the XMLCommetns to the ILua.TYPES enum values to avoid the warnings and provide context for developers
